src/nvim/eval/funcs.c

```diff
@@ -2663,7 +2663,7 @@ static void f_fnamemodify(typval_T *argvars, typval_T *rettv, FunPtr fptr)
   char buf[NUMBUFLEN];
   const char *fname = tv_get_string_chk(&argvars[0]);
   const char *const mods = tv_get_string_buf_chk(&argvars[1], buf);
-  if (fname == NULL) {
+  if (mods == NULL || fname == NULL) {
     fname = NULL;
   } else if (mods != NULL && *mods != NUL) {
```

Comment:
This line and the `*mods != NUL` line below should also need change, otherwise `len` is still zero when `*mods` is NULL.

Reply:
Thanks, I missed that change -- and now #16367 is _actually_ fixed :)